File: src/languages-meta/grammar/grammar.grammar.model.ts
export class Grammar {
    constructor(
        public name: Name,
        public rules: Rule[],
        public prattRules: PrattRule[]
    ) {}
}

export class Rule {
    constructor(
        public name: Name,
        public annotation: RuleAnnotation | undefined,
        public versionAnnotations: VersionAnnotation[],
        public body: RuleBody
    ) {}
}

export enum RuleAnnotation {
    NoSkip,
    Atomic
}

export class PrattRule {
    constructor(
        public name: Name,
        public versionAnnotations: VersionAnnotation[],
        public operator: PrattOperator,
        public primary: PrattPrimary
    ) {}
}

export class PrattOperator {
    constructor(
        public type: PrattOperatorType,
        public name: Name,
        public versionAnnotations: VersionAnnotation[],
        public body: RuleBody
    ) {}
}

export class PrattPrimary {
    constructor(
        public name: Name,
        public body: RuleBody
    ) {}
}

export enum PrattOperatorType {
    Prefix,
    Postfix,
    Left,
    Right
}

export class VersionAnnotation {
    constructor(
        public type: VersionAnnotationType,
        public version: VersionNumber
    ) {}
}

export enum VersionAnnotationType {
    Enabled,
    Disabled
}

export type VersionNumber = VersionSegment[];

export type VersionSegment = string;

export type RuleBody = ChoiceRule | SequenceRule;

export class ChoiceRule {
    constructor(public choices: SequenceRule[]) {}
}

export class SequenceRule {
    constructor(public element: RuleElement) {}
}

export type RuleElement = CountedRuleElement | NegativeLookahead;

export class CountedRuleElement {
    constructor(
        public label: Label | undefined,
        public countableRuleElement: CountableRuleElement,
        public count: Count | undefined,
        public versionAnnotations: VersionAnnotation[]
    ) {}
}

export type CountableRuleElement = RuleReference | StringElement | CharSet | AnyElement | RuleBody;

export enum Count {
    OneOrMore,
    ZeroOrMore,
    Optional
}

export type Label = Name;

export type Name = Identifier;

export class RuleReference {
    constructor(public names: Name[]) {}
}

export class StringElement {
    constructor(public value: string) {}
}

export class CharSet {
    constructor(
        public negated: boolean,
        public startChar: CharSetChar,
        public endChar: CharSetChar | undefined
    ) {}
}

export type CharSetChar = string;

export class AnyElement {}

export class NegativeLookahead {
    constructor(public content: CharSet | StringElement) {}
}

export type Identifier = string;

export type Trivia = LineComment | BlockComment | Whitespace;

export class LineComment {
    constructor(public value: string) {}
}

export class BlockComment {
    constructor(public value: string) {}
}

export class Whitespace {
    constructor(public value: string) {}
}

File: src/languages-meta/grammar/grammar.grammar.model.rs
pub struct Grammar {
    pub name: Name,
    pub rules: Vec<Rule>,
    pub prattRules: Vec<PrattRule>,
}

pub struct Rule {
    pub name: Name,
    pub annotation: Option<RuleAnnotation>,
    pub versionAnnotations: Vec<VersionAnnotation>,
    pub body: RuleBody,
}

pub enum RuleAnnotation {    NoSkip,    Atomic,}

pub struct PrattRule {
    pub name: Name,
    pub versionAnnotations: Vec<VersionAnnotation>,
    pub operators: Vec<PrattOperator>,
    pub primary: PrattPrimary,
}

pub struct PrattOperator {
    pub type: PrattOperatorType,
    pub name: Name,
    pub versionAnnotations: Vec<VersionAnnotation>,
    pub body: RuleBody,
}

pub struct PrattPrimary {
    pub name: Name,
    pub body: RuleBody,
}

pub enum PrattOperatorType {    Prefix,    Postfix,    Left,    Right,}

pub struct VersionAnnotation {
    pub type: VersionAnnotationType,
    pub version: VersionNumber,
}

pub enum VersionAnnotationType {    Enabled,    Disabled,}

pub type VersionNumber =Vec<VersionSegment>;

pub type VersionSegment =String;

pub type RuleBody =ChoiceRule | SequenceRule;

pub struct ChoiceRule {
    pub choices: Vec<SequenceRule>,
}

pub struct SequenceRule {
    pub elements: Vec<RuleElement>,
}

pub type RuleElement =CountedRuleElement | NegativeLookahead;

pub struct CountedRuleElement {
    pub label: Option<Label>,
    pub countableRuleElement: CountableRuleElement,
    pub count: Option<Count>,
    pub versionAnnotations: Vec<VersionAnnotation>,
}

pub type CountableRuleElement =RuleReference | StringElement | CharSet | AnyElement | RuleBody;

pub enum Count {    OneOrMore,    ZeroOrMore,    Optional,}

pub type Label =Name;

pub type Name =Identifier;

pub struct RuleReference {
    pub names: Vec<Name>,
}

pub struct StringElement {
    pub value: String,
}

pub struct CharSet {
    pub negated: boolean,
    pub startChars: Vec<CharSetChar>,
    pub endChars: Vec<Option<CharSetChar>>,
}

pub type CharSetChar =String;

pub struct AnyElement {
}

pub struct NegativeLookahead {
    pub content: CharSet | StringElement,
}

pub type Identifier =String;

pub type Trivia =LineComment | BlockComment | Whitespace;

pub struct LineComment {
    pub value: String,
}

pub struct BlockComment {
    pub value: String,
}

pub struct Whitespace {
    pub value: String,
}